src/outputs/output_utils.hpp

```diff
@@ -41,31 +46,89 @@ namespace parthenon {
 namespace OutputUtils {
 // Helper struct containing some information about a variable
 struct VarInfo {
+ public:
+  static constexpr int VNDIM = MAX_VARIABLE_DIMENSION;
   std::string label;
   int num_components;
-  int nx6;
-  int nx5;
-  int nx4;
-  int nx3;
-  int nx2;
-  int nx1;
   int tensor_rank; // 0- to 3-D for cell-centered variables, 0- to 6-D for arbitrary shape
                    // variables
   MetadataFlag where;
   bool is_sparse;
   bool is_vector;
+  IndexShape cellbounds;
   std::vector<std::string> component_labels;
-  int Size() const { return nx6 * nx5 * nx4 * nx3 * nx2 * nx1; }
+  std::vector<TopologicalElement> topological_elements;
+  int ntop_elems;
+  bool element_matters;
+
+  Triple_t<int> GetNKJI(const IndexDomain domain) const;
+  Triple_t<IndexRange> GetPaddedBoundsKJI(const IndexDomain domain) const;
+
+  int Size() const;
+  // Includes topological element shape
+  int TensorSize() const;
+  // Size of region that needs to be filled with 0s if not allocated
+  int FillSize(const IndexDomain domain) const;
+  // number of elements of data that describe variable shape
+  int GetNDim() const;
+
+  template <typename T>
+  int FillShape(const IndexDomain domain, T *data) const {
+    if (where == MetadataFlag({Metadata::None})) {
```

Comment:
Broadly, `Metadata::None` is a special snowflake everywhere. I couldn't figure out how to not special-case it. But most of the special-casing is hidden.